### ISSUE TITLE
Engine should create output writers/streams on behalf of tools that require them

### ISSUE THREAD

**Issue by droazen** [broadinstitute/gatk#121]

**Comment by akiezun** [comment]
to clarify: the requirement is to implement a new annotation for outputs, mark output fields with that annotation, then add code to open the writers to the engine (probably in GATKTool) and close the writers on exit (regardless of fail/success).

**Comment by lbergelson** [comment]
I'm not sure how much of this we want to tackle before dataflowing. We are
probably going to want to provide a standardized output format that will
map nicely to Dataflow. Currently this means a single output pcollection,
although some tools obviously will need multiple output streams.

**Comment by droazen** [comment]
Yes, we shouldn't assume this will necessarily be done any particular way (eg., through annotations) unless we have more working examples of dataflow tools from which to generalize.

**Comment by akiezun** [comment]
we need this independently of dataflow (needs to work one file walkers as well as dataflow)

**Comment by akiezun** [comment]
alpha candidate?

**Comment by droazen** [comment]
yes!

**Comment by droazen** [comment]
This applies to spark too.

**Comment by droazen** [comment]
In https://github.com/broadinstitute/gatk/pull/1146 we've at least centralized SAM/BAM writer creation methods in the superclasses -- this is good enough for alpha. In beta we can add the fancy auto-management of outputs we are dreaming about.

**Comment by cmnbroad** [comment]
There is a [branch/commit](https://github.com/broadinstitute/gatk/commit/2fb990dfeb7f5be20d6d09e184327d3872094f26) here with changes to centralize creation of the output VCFWriter for all (non-Spark) tools in order to centralize index creation, which is an incremental step in this direction.

**Comment by droazen** [comment]
@cmnbroad If you have any ideas about how this should be done, feel free to suggest an approach

**Comment by lbergelson** [comment]
What happened with that branch?

**Comment by droazen** [comment]
It is blocked by another branch currently under review, I believe.
